DEVX-1292: Producer/consumer example: Spring Boot (#428)
confluentinc/examples: clients/cloud/java-springboot/README.adoc

. Verify your Confluent Cloud Schema Registry credentials work from your host. 
In the output below, substitute your values for `<SR API KEY>`, `<SR API SECRET>`, and `<SR ENDPOINT>`.
+
[source,shell]
.View the list of registered subjects
----
$ curl -u <SR API KEY>:<SR API SECRET> https://<SR ENDPOINT>/subjects
----

This Spring Boot application has two components - Producer (`ProducerExample.java`) and Consumer (`ConsumerExample.java`).
Both components will be initialized during the Spring Boot application startup.
The producer writes Kafka data to a topic in Confluent Cloud.
Each record has a String key representing a username (e.g. `alice`) and a value of a count, formatted as Avro object

[source,json]
----
{"namespace": "io.confluent.examples.clients.cloud",
 "type": "record",
 "name": "DataRecordAvro",
 "fields": [
     {"name": "count", "type": "long"}
 ]
}
----

> Run producer and consumer application by executing `startProducerConsumer.sh`.

Reviewer: ```suggestion
Run producer and consumer application by executing `./startProducerConsumer.sh`.
```

Author: done

> Run producer and consumer application by executing `startProducerConsumer.sh`.

Reviewer 2: ```suggestion
Run producer and consumer application by executing:
[source,shell]
----
./startProducerConsumer.sh
----
```

2020-02-13 14:41:58.249  INFO 44191 --- [ntainer#0-0-C-1] i.c.e.c.c.springboot.ConsumerExample     : received alice {"count": 4}
2020-02-13 14:41:58.249  INFO 44191 --- [ntainer#0-0-C-1] i.c.e.c.c.springboot.ConsumerExample     : received alice {"count": 5}
2020-02-13 14:41:58.249  INFO 44191 --- [ntainer#0-0-C-1] i.c.e.c.c.springboot.ConsumerExample     : received alice {"count": 6}
2020-02-13 14:41:58.249  INFO 44191 --- [ntainer#0-0-C-1] i.c.e.c.c.springboot.ConsumerExample     : received alice {"count": 7}
2020-02-13 14:41:58.249  INFO 44191 --- [ntainer#0-0-C-1] i.c.e.c.c.springboot.ConsumerExample     : received alice {"count": 8}
2020-02-13 14:41:58.249  INFO 44191 --- [ntainer#0-0-C-1] i.c.e.c.c.springboot.ConsumerExample     : received alice {"count": 9}
----

NOTE: When you are done, press kbd:[Ctrl + c].

== Example 2: Kafka Streams with Spring Boot

The Kafka Streams API reads the same topic from Confluent Cloud and does a stateful sum aggregation, also a rolling sum of the counts as it processes each record.

.Run the Kafka Streams application.
[source,shell]
----

> include::startStreams.sh[lines=28]

Reviewer: @gAmUssA in GH this doesn't render well.  Should it be just `./startStreams.sh`?

Author: done

> include::startStreams.sh[lines=28]

Reviewer 2: This shell command looks incorrect to me, can you validate that's what it's supposed to be?

Author: @rspurgeon update your copy

Author: this command is gone now

= Spring Boot with Java Example Confluent Cloud Client

> NOTE: Produce messages to and consume messages from [Confluent Cloud](https://www.confluent.io/confluent-cloud/?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.clients-ccloud) using the Spring Boot using Java API.

Reviewer 2: ```suggestion
NOTE: Produce and consume messages from [Confluent Cloud](https://www.confluent.io/confluent-cloud/?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.clients-ccloud) using the Spring Boot Java Framework.
```

== Prerequisites

> * Java 1.8 or higher to run the demo application

Reviewer 2: ```suggestion
* Java 1.8
```

Follow link:https://github.com/confluentinc/configuration-templates/tree/master/README.md[these detailed instructions] to properly create this file.
* If you are running on Confluent Cloud, you must have access to a link:https://www.confluent.io/confluent-cloud/?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.clients-ccloud[Confluent Cloud] cluster.

[source]
.Confluent Cloud config file example
----
$ cat $HOME/.ccloud/config
bootstrap.servers=<BROKER ENDPOINT>
ssl.endpoint.identification.algorithm=https
security.protocol=SASL_SSL
sasl.mechanism=PLAIN
sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username\="<API KEY>" password\="<API SECRET>";
schema.registry.url=<SR ENDPOINT> 
basic.auth.credentials.source=USER_INFO 
schema.registry.basic.auth.user.info=<SR_KEY:SR_PASSWORD>
----

> NOTE: To building and execute all examples, run `start.sh`.
> Run `stop.sh` to stop both examples.

Reviewer 2: ```suggestion
NOTE: To build and execute all examples run `start.sh` and use `stop.sh` to cleanup.
```

== Example 1: Producing and Consuming Avro messages

> This example is uses values that formatted as Avro and integrates with the Confluent Cloud Schema Registry.

Reviewer 2: ```suggestion
This example uses values that are formatted in Avro and integrates with the Confluent Cloud Schema Registry.
```

> Before using Confluent Cloud Schema Registry, check its [availability and limits](https://docs.confluent.io/current/cloud/limits.html?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.clients-ccloud).
> Note that your VPC must be able to connect to the Confluent Cloud Schema Registry public internet endpoint.

Reviewer 2: ```suggestion
Before using Confluent Cloud Schema Registry, check [availability and limits](https://docs.confluent.io/current/cloud/limits.html?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.clients-ccloud), and be sure you have connectivity to the Confluent Cloud Schema Registry public endpoint.
```

= Spring Boot with Java Example Confluent Cloud Client

NOTE: Produce messages to and consume messages from [Confluent Cloud](https://www.confluent.io/confluent-cloud/?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.clients-ccloud) using the Spring Boot using Java API.

> == Prerequisites

Reviewer: @gAmUssA unfortunately the section Prerequisites format has changed since this PR was opened.  Can you please make it similar to https://github.com/confluentinc/examples/blob/5.4.1-post/clients/cloud/groovy/README.adoc ?

Author: updated

> * Java 1.8 or higher to run the demo application

Reviewer: `gradle` is a prereq

Author: we use Gradle wrapper. Gradle isn't prerequisite in this case, because wrapper will download required version of gradle

* Access to a [Confluent Cloud](https://www.confluent.io/confluent-cloud/?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.clients-ccloud) cluster

> * Initialize a properties file at `src/main/resources/application-ccloud.properties` with configuration to your Confluent Cloud cluster:

Reviewer: @gAmUssA all the other examples use a standard Java properties file with just the key properties to connect to a cluster.  

```
$ cat $HOME/.ccloud/config
bootstrap.servers=<BROKER ENDPOINT>
ssl.endpoint.identification.algorithm=https
security.protocol=SASL_SSL
sasl.mechanism=PLAIN
sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username\="<API KEY>" password\="<API SECRET>";
```

However, this PR deviates from that.  Can this example follow the other examples?  If Spring needs `.application-ccloud.properties`, then the example should auto-generate it from the standard Java properties file.

Reasons to do it this way:

1. Consistency
2. Eventually Confluent CLI will autogenerate the config file, so this example can just consume it.

Author: it's done!

[source,shell]
.application-ccloud.properties
----

> ❯ cat src/main/resources/application-ccloud.properties

Reviewer: @gAmUssA all client examples use a common config file as describe in https://github.com/confluentinc/configuration-templates/blob/master/README.md

Which means that this `application-ccloud.properties` file must be auto-generated from that common config file.  The value in doing it this way:

- provides consistency between examples
- enables users to try different code languages using the same config file
- enables us to test

# topic config
io.confluent.developer.config.topic.name=test
io.confluent.developer.config.topic.replicas=3
io.confluent.developer.config.topic.partitions=6

# common configs
spring.kafka.properties.ssl.endpoint.identification.algorithm=https
spring.kafka.properties.sasl.mechanism=PLAIN
spring.kafka.properties.request.timeout.ms=20000
spring.kafka.properties.bootstrap.servers=bootstrap-server.confluent.cloud:9092
spring.kafka.properties.retry.backoff.ms=500
spring.kafka.properties.sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username="<API_KEY>" password="<API_SECRET>";
spring.kafka.properties.security.protocol=SASL_SSL

> # Cloud SR Config

Reviewer: There should be two separate examples: 

1. One without Avro/SR
2. One with Avro/SR

Author: I would submit separate PR for JSON version of this example.

spring.kafka.properties.basic.auth.credentials.source=USER_INFO
spring.kafka.properties.schema.registry.basic.auth.user.info=SR_KEY:SR_SECRET
spring.kafka.properties.schema.registry.url=https://sr.confluent.cloud

# producer configuration
spring.kafka.producer.key-serializer=org.apache.kafka.common.serialization.StringSerializer
spring.kafka.producer.value-serializer=io.confluent.kafka.serializers.KafkaAvroSerializer

# consumer configuration
spring.kafka.consumer.group-id=java-springboot
spring.kafka.consumer.key-deserializer=org.apache.kafka.common.serialization.StringDeserializer
spring.kafka.consumer.value-deserializer=io.confluent.kafka.serializers.KafkaAvroDeserializer
----

> == Hello World!

Reviewer: @gAmUssA for consistency, please use the same formatting as described in another example, e.g. the Java one

Author: same formatting for configuration file or README?

Reviewer: Should be both: -- same configuration file can be used across all examples, and README so that the text/example is the same

This Spring Boot application has two components - Producer (`ProducerExample.java`) and Consumer (`ConsumerExample.java`).
Both components will be initialized during the Spring Boot application startup.
The producer writes Kafka data to a topic in Confluent Cloud.
Each record has a String key representing a username (e.g. `alice`) and a value of a count, formatted as Avro object
+
[source,json]
----
{"namespace": "io.confluent.examples.clients.cloud",
 "type": "record",
 "name": "DataRecordAvro",
 "fields": [
     {"name": "count", "type": "long"}
 ]
}
----

The consumer reads the same topic from Confluent Cloud and prints data to the console.

The Kafka Streams API reads the same topic from Confluent Cloud and does a stateful sum aggregation, also a rolling sum of the counts as it processes each record.

. Run the producer, passing in arguments for (a) the local file with configuration parameters to connect to your Confluent Cloud instance and (b) the topic name:

+
[source, shell]
----
 # Build the application
 $ ./gradlew clean build
   
 # Run the app

>  $ java -cp build/libs/java-springboot-0.0.1-SNAPSHOT.jar -Dloader.main=io.confluent.examples.clients.cloud.springboot.kafka.SpringbootKafkaApplication org.springframework.boot.loader.PropertiesLauncher <1>

Reviewer: `<1>` (and numbers below) -- what is their purpose?  It doesn't work with them so they should be removed
